quickwit/quickwit-storage/src/opendal_storage/google_cloud_storage.rs

```diff
@@ -0,0 +1,121 @@
+//
+// AGPL:
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program. If not, see <http://www.gnu.org/licenses/>.
+
+use std::path::PathBuf;
+use std::sync::Arc;
+
+use async_trait::async_trait;
+use once_cell::sync::OnceCell;
+use quickwit_common::uri::Uri;
+use quickwit_config::{GoogleCloudStorageConfig, StorageBackend};
+use regex::Regex;
+
+use super::OpendalStorage;
+use crate::debouncer::DebouncedStorage;
+use crate::{Storage, StorageFactory, StorageResolverError};
+
+/// Google cloud storage resolver.
+pub struct GoogleCloudStorageFactory {
+    storage_config: GoogleCloudStorageConfig,
+}
+
+#[async_trait]
+impl StorageFactory for GoogleCloudStorageFactory {
+    fn backend(&self) -> StorageBackend {
+        StorageBackend::Google
+    }
+
+    async fn resolve(&self, uri: &Uri) -> Result<Arc<dyn Storage>, StorageResolverError> {
+        let storage = from_uri(&self.storage_config, uri)?;
+        Ok(Arc::new(DebouncedStorage::new(storage)))
+    }
+}
+
+fn from_uri(
+    google_cloud_storage_config: &GoogleCloudStorageConfig,
+    uri: &Uri,
+) -> Result<OpendalStorage, StorageResolverError> {
+    let credential_path = google_cloud_storage_config
+        .resolve_credential_path()
+        .ok_or_else(|| {
+            let message = format!(
+                "could not find Google credential path in environment variable `{}` or storage \
+                 config",
+                GoogleCloudStorageConfig::GOOGLE_CLOUD_STORAGE_CREDENTIAL_PATH_ENV_VAR
+            );
+            StorageResolverError::InvalidConfig(message)
+        })?;
+    let (bucket_name, prefix) = parse_google_uri(uri).ok_or_else(|| {
+        let message = format!("failed to extract bucket name from google URI: {uri}");
+        StorageResolverError::InvalidUri(message)
+    })?;
+
+    let mut cfg = opendal::services::Gcs::default();
+    cfg.credential_path(&credential_path);
+    cfg.bucket(&bucket_name);
+    cfg.root(&prefix.to_string_lossy());
+
+    let store = OpendalStorage::new_google_cloud_storage(uri.clone(), cfg)?;
+    Ok(store)
+}
+
+/// TODO: maybe we can also support `gs://bucket`?
+fn parse_google_uri(uri: &Uri) -> Option<(String, PathBuf)> {
+    // Ex: google://bucket/prefix.
+    static URI_PTN: OnceCell<Regex> = OnceCell::new();
+
+    let captures = URI_PTN
+        .get_or_init(|| {
+            Regex::new(r"google(\+[^:]+)?://(?P<bucket>[^/]+)(/(?P<prefix>.+))?")
```

Comment:
I think `gs` is the usual protocol for google cloud storage.
Have you seen google:// somewhere?

Reply:
I think your regex does not allow `google://bucket/`.
Shouldn't we allow it?

Reply:
Those logic copied from azure and check by `test_parse_google_uri`. `google://bucket/` should work.

Reply:
> I think `gs` is the usual protocol for google cloud storage.

Addressed.